### ISSUE TITLE
Question about ELBO

### ISSUE THREAD

**Issue by won-bae** [Jianf-Wang/NP-Match#1]
Hi Jianfeng,

Thank you for making the code public for awesome work. While reading your paper, I had a hard time to understand two parts of the paper. It would be highly appreciated if you clarify them. 

1. As far as I understood, the latent path takes the features and corresponding labels (both real and pseudo) as inputs. But, when the model makes predictions on target points in training time, isn't there information leak especially for labeled data? I am not sure how the latent path and decoder learns something meaningful when real labels are passed as inputs.
2. Given that the constant term in ELBO is $E_q [ \log p(y_{1:m})]$ where $y_{1:m}$ does not depend on $x_{1:m}$, I guess your graphical model looks the same as Figure2d of Neural Processes paper (https://arxiv.org/abs/1807.01622), and my previous question is related to this. 
It makes sense to use real labels as inputs "only if" the context sets and target sets are not overlapping, which is not the case in your problem setting. As the labeled data points are overlapping in the context and target sets, it seems weird to have two different dependency assumption between $y_{1:m}$ and $x_{1:m}$. In the context set, you assumed $y_{1:m}$ are independent to $x_{1:m}$ but for the same variables in the target set, you assume $y_{1:m}$ (or $y_{m+1:m+r}$) depends on both $z$ and $x_{1:m}$ (or $x_{m+1:m+r}$).   

I feel like I am missing something important here. Any guidance would be appreciated. Thank you!

**Comment by Jianf-Wang** [comment]
Hi, thanks for your attention to this work.

In fact, we also considered this before. Theoretically, if we strictly follow the original NPs to ensure that context sets and the target sets are not overlapping, the information leak and the two different dependency problems for labelled data won’t be there. To achieve this in semi-supervised learning (SSL), a better and more natural way seems to only use labelled data as context and to only use unlabelled data as target, since original NPs are designed to make predictions for target points conditioned on context points. However, this is not achievable directly, since SSL is a new task compared to the tasks in the original NPs paper, and this new task has some different characteristics.

A key characteristic of SSL is that we do not have labels for unlabelled data. In recent SOTA deterministic methods for SSL (such as FixMatch), a common way is to train the model on labelled data at first, and then use the model to predict and select pseudo-labels for unlabelled data. When adapting NPs to SSL, if we only use unlabelled data as target points, we should consider how to get pseudo-labels (the overall framework is randomly initialized) in the early stage of training. As this is the first attempt to apply NPs to SSL, we choose a naïve way, namely using labelled data as target points as well, since we also need to make predictions for labelled data and train the model with them. Such naïve way is also consistent with the SOTA SSL methods, so that the performance is guaranteed. As the training goes, the overall framework becomes better, and it can select more and more unlabelled data. Since unlabelled data are much more than labelled data, there will be much more unlabelled data in each batch (unlabelled : labelled = 7:1 in CIFAR and STL-10). When we use the mean aggregator, the output will be similar to the case that we only use unlabelled data as target points, and unlabelled data also dominant  the gradient descent direction.
Based on these, we decided to follow the original NPs paper to write the ELBO and the theory part. As a result, there is a gap between our real implementation and the theory.  We consider that such gap does not matter too much when unlabelled data are much more than the labelled data, but when their numbers are close, such gap and overlapping may affect the performance, which could be observed from our Imagenet result. For the ImageNet result, we can only choose to set unlabelled : labelled = 1:1, due to the limited GPU resources, and the performance is not quite good.

We did not clarify this gap in the paper, which may cause confusion. I hope this explanation is helpful. The aforementioned naïve  way may not be a very good choice, but at least can be used under SSL setting to get comparable results.

**Comment by won-bae** [comment]
Hi Jianfeng,

Thank you for the clarification. Yeah, it would've been better if there was a clarification for this in the paper. But, now it makes sense that in practice, it may not be a critical problem with more unlabeled data. I guess it would be an interesting future work to come up with an algorithm that is more aligned with theory.
Thank you for sharing your great work again!
